docs/contributing/Nullable Annotations.md

```diff
@@ -0,0 +1,74 @@
+# Nullable Annotations
+
+This document describes how nullable annotations should be approached in the 
+Roslyn code base. The default is to simply follow [the same guidance](https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/api-guidelines/nullability.md)
+as the [dotnet/runtime](github.com/dotnet/runtime) repository. This document
+serves to detail the places where the guidance differs for Roslyn and 
+re-emphasize rules that come up frequently.
+
+## Annotation Guidance 
+
+- **DO** annotate all new code
+- **DO** use the null suppression operator in places where a method body has
+ensured a value cannot be `null` but the compiler is unable to track it. 
+Example:
+```cs
+struct SyntaxTrivia {
+    void Example() {    
+        if (this.HasStructure) {
+            Use(this.GetStructure()!);
+        }
+    }
+}
+```
+- **DO** use explicit validation such as `RoslynDebug.Assert` or 
+`Contract.ThrowIfNull` to validate state that is passed into the method if it 
+avoids the need for suppressions later in the method.
+```cs
+void M1(SyntaxNode node) {
+    RoslynDebug.Assert(node.Parent is object);
+    ...
+    M2(node.Parent);
+}
+```
+- **AVOID** annotating and refactoring code in the same commit. This can
+significantly increase the cost of reviewing. Prefer instead to separate 
+annotations and significant refactorings into separate commits. Small changes,
+and modernizing style (change `default(SyntaxToken)` to `default`) are
+acceptable.
+- **DO NOT** be concerned with the validity of annotations after an object has 
+been disposed or otherwise entered into an invalid state. For example
+manipulating a pooled object after it's been freed. This is a violation of the 
+objects contract and annotations aren't meaningful here.
+- **DO NOT** be concerned with the validity of an enumerator's `Current` 
+annotations when used before `MoveNext` has been called or after `MoveNext` has 
+returned false. As with `Dispose`, such use is considered invalid, and we don't
+want to harm the correct consumption of `Current` by catering to incorrect
+consumption.
+- **DO** use the adjective `Require` when defining a member that exist only
+to provide a non-null returning version of a method that returns a nullable 
+reference type. For example: `GetRequiredSemanticModel` model returns a non-null 
+value from `GetSemanticModel`
```

Comment:
Is this for internal methods only or future public APIs that parallel legacy null-returning API? The IDE was doing this with internal methods but we were somewhat hedging on whether we wanted to make those public. We borrowed the name from some methods (GetRequiredLanguageService) that are very old and so on one hand liked the consistency, but weren't sure if that was following a consistency that wasn't really intended at the time. Those methods predate Roslyn 1.0 and thus establishing a naming convention for methods wasn't in the purview of that!

Did the framework do anything here?